docs/rfcs/017-pageserver-op-atomicity.md

```diff
@@ -0,0 +1,153 @@
+
+Below is a brief description of all the pageserver tenant/timeline
+operations, and how the steps of creating/deleting local files or
+directories and uploads to S3 are performed. The steps are listed in
+such an order that each operation can be sanely recovered or aborted,
+if the pageserver crashes while the operation is being perfromed.
+
+## Create tenant
+
+Create an empty tenant. It doesn't have any timelines initially.
+
+1. Create local tenant-directory with .temp extension
+2. Create tenant.json file in the directory, with a special flag
+   indicating that the tenant-creation is in progress
+3. Rename the local tenant directory in place
+4. Upload the tenant.json file to S3, without the flag
+5. Update the local file, removing the flag
+
+At pageserver startup, if we see a tenant.json file with the special
+flag, check if the tenant exists in S3. If not, remove the local directory.
+Otherwise remove the flag from local file.
+
+## Create timeline
+
+Create a timeline for a tenant, as result of running initdb.
+
+1. create timeline directory locally, with .temp extension
+2. run initdb, creating the initial set of layers
+3. upload all layer files to S3
+4. upload metadata file to S3
+5. update tenant.json file in S3
+6. Rename local directory in place
+
+If we crash before step 5, S3 may have a timeline metadata file and some
+layer files, without corresponding entry in tenant.json file. That's OK.
+Whenever we see that, we can delete the leftover timeline files.
+
+If we want to make that less scary, we could update a tenant.json file in S3
+twice. First, add the new timeline ID to the file with a flag indicating
+that it's being created. Do that before uploading anything else to S3. And
+then in step 5, update tenant.json to indicate that the creation is complete.
+
+## Branch timeline
+
+Create a new timeline with an existing timeline as parent
+
+1. create timeline directory locally, with .temp extension
+2. create metadata file in the local directory
+3. upload metadata file to S3
+4. update tenant.json file in S3
+5. Rename local directory in place
+
+Like with Create timeline, if we crash between steps 3 and 4, we will
+leave behind a timeline metadata file with no corresponding entry in
+tenant.json.  That's harmless.
+
+## Delete timeline
+
+1. rename local timeline directory to have .temp extension
+2. Update tenant.json file in S3
+3. delete index file from S3
+4. delete layer files from S3
+5. delete local directory
+
+Like with creation, if this is interrupted, we will leave behind
+timeline files in S3 with no corresponding entry in tenant.json. If we
+want to make that less scary, we can update tenant.json in step 2 with
+a tombstone flag for the timeline we're removing, instead of removing
+the entry for it outright.
+
+## Delete tenant
+
+1. rename local tenant directory to have .temp extension
+2. delete tenant.json file in S3
+3. delete all timeline index files from S3
```

Comment:
Again, the same scheduling-vs-waiting question appears here: we don't care much about the stray timelines in S3, so might it be ok to schedule such deletions in background and go on with local directory deletion?

Either way, if pageserver crashes, we won't delete any files, but with the scheduling, we have a chance to retry later and not to wait on extra network requests.

Reply:
From my point of view synchronization between pageserver and S3 can be done quite easily:
1. The main source of truth is pageserver. It means that data is first updated locally at pageserver and them sometimes uploaded to S3.  Each layer data and metadata (index) file have associated LSN, There can be some background process (like checkpointer) which uploads all layer with LSN <= remote consistency threshold.
2. All data files (i.e. layers) are uploaded before metadata. Once his upload process complete uploading of all tenant's files with LSN <= upload horizon, it updates `tenant.json`, storing new `remote-consistent-lsn` in it.
3. When tenant is accessed first time, pageserver  dowloads  `tenant.json` and so determine which information is available at S3. Pageserver should never have older versions of files than S3. But it may miss this tenent at all (it can happen in case of page server corruption otr migration of tenant to the new pageserver). In this case timeline files should be downloaded fro s3 on demand.

Reply:
Looks reasonable, pageserver seems to follow these proposals partially already, but some details might need more discussion.


1.  > The main source of truth is pageserver. It means that data is first updated locally at pageserver and them sometimes uploaded to S3

Changes to layers after freezing, compaction and gc are directly submitted for synchronization with S3 at its current implementation, neither the [recently proposed one](https://github.com/neondatabase/neon/pull/2231) alters this generally.

Yet, we delete files from S3 with a relative caution now, only during compaction or GC, so we can always redownload some extra layers after the crash. Should we remove such files eagerly, based on pageserver's local FS contents?
We attempt to synchronize pageservers' writes into S3 with different prefixes in S3 and `detach` and `attach` per tenant, run in correct order, but, AFAIK, during tenant relocation there's a possibility for both pageservers to write into S3 for some short time, @LizardWizzard was implementing the cloud part and can correct me here.

If it's true and we ever allow multiple pageservers to write into the same S3 timeline simultaneously, deletions might be quite harmful.

2. >  it updates tenant.json, storing new remote-consistent-lsn in it.

While now we attach/detach by tenant, we used to do that by tenant and could return to that when download on demand works well.
In that case, multiple pageservers *will* write into the same file and we could loose the information. 
We currently consider `remote-consistent-lsn` as timeline's data and write into its json files, along with `metadata` bytes and a new layer file path added (this way, we avoid S3 list when actually listing S3 timeline contents): we have to do some of those writes anyway, so we do both into the same file.

I would keep `tenant.json` smaller and less volatile, if possible: with download on-demand implemented, we might be tempted to get back to attach/detach per timeline and that would mean more pageservers would write into the same json file that they need to reread and ensure no one writes into that.
Easier to do that without writing frequently changing `remote-consistent-lsn` from multiple pageservers: we already ensure attach/detach happens in a certain order.

It's interesting to note, that no one considers `latest_gc_cutoff_lsn` from `metadata` neither in this RFC nor in the current code and the new PR, so seems that we're fine with recalculating it?
If we are, and we're never updating `disk_consistent_lsn` in `metadata` without layer file created, we could eliminate almost all json file updates on S3.